src/thermo/Species.cpp

```diff
@@ -29,12 +32,54 @@ Species::Species(const std::string& name_, const compositionMap& comp_,
     , charge(charge_)
     , size(size_)
 {
+    double wt = 0.;
+    const map<string, double>& elements = elementSymbolToWeight();
+    const map<string, double>& isotopes = isotopeSymbolToWeight();
+    for (const auto& comp : comp_) {
+        auto search = elements.find(comp.first);
+        if (search != elements.end() && search->second > 0) {
+            wt += search->second * comp.second;
+        } else {
+            search = isotopes.find(comp.first);
+            if (search != isotopes.end() && search->second > 0) {
+                wt += search->second * comp.second;
+            } else {
+                m_unknownElements.push_back(comp.first);
+            }
+        }
+    }
+    if (m_unknownElements.empty()) {
+        setMolecularWeight(wt);
+    }
 }
 
 Species::~Species()
 {
 }
 
+double Species::molecularWeight() {
+    if (!m_unknownElements.empty()) {
+        throw CanteraError(
```

Comment:
> Can we make the calculation of the molecular weight lazy, and do it in this method rather than the constructor?

Sure, I think this makes a lot of sense. One minor concern that is probably already present, what happens in this case:

```python
s = ct.Species(name='ARstandard', {"AR": 1})
s.molecular_weight
39.95
gas = ct.Solution('test/data/ideal-gas.yaml', 'element-remote')
gas.add_species(s)  # this is made up, I don't remember the API to do this off-hand, but I think this is possible
s.molecular_weight
????
```
The only reason to check if the value is `Undef` in the setter is to save some unnecessary computation... on the other hand, maybe this computation isn't that intensive, and it's only done once, so it should always be done when a species is added to a phase. What do you think?

Reply:
I suppose there are two options here, both of which start with `Phase::addSpecies` always setting the molecular weight of the `Species` object.

1. `Species::setMolecularWeight` just accepts the updated value, so it now returns 38.
2. `Species::setMolecularWeight` throws if the stored value is not either `Undef` or very near the input value, which would imply that the weight had already been calculated based standard element definitions and accessed elsewhere, leading to potential inconsistency.

I think the second option is better, and closer to what you already have.